new-cdc/eventuate-local-java-cdc-connector-mysql-binlog/src/main/java/io/eventuate/local/mysql/binlog/MySqlBinaryLogClient.java

```diff
@@ -232,19 +293,25 @@ else if (tableMapEventByTableId.containsKey(eventData.getTableId())) {
 
       SchemaAndTable schemaAndTable = new SchemaAndTable(tableMapEventData.getDatabase(), tableMapEventData.getTable());
 
-      BinlogEntry entry = extractor.extract(schemaAndTable, eventData, binlogFilename, offset);
+      BinlogEntry entry = extractor.extract(schemaAndTable, eventData, binlogFilename, offset, lastGTID);
 
-
-      if (!shouldSkipEntry(startingBinlogFileOffset, entry.getBinlogFileOffset())) {
+      if ((useGTIDsWhenPossible && isGTIDSupported()) || !shouldSkipEntry(startingBinlogFileOffset, entry.getBinlogFileOffset())) {
         binlogEntryHandlers
               .stream()
               .filter(bh -> bh.isFor(schemaAndTable))
-              .forEach(binlogEntryHandler -> binlogEntryHandler.publish(entry));
+              .forEach(binlogEntryHandler -> {
+                offset = extractOffset(event);
+                BinlogFileOffset binlogFileOffset = new BinlogFileOffset(binlogFilename, offset, lastGTID);
+
+                cdcDataPublisherTransactionTemplate.inTransaction(() -> {
```

Comment:
This isn't correct.
`inTransaction()` should wrap ` binlogEntryHandlers.stream()` so that the all messages are published and the `binlogFileOffset` is published within a single transaction .